### PR TITLE
feat: User Agent Enhancements Part 2 (datastore, graphql, pubsub cross-communication)

### PR DESCRIPTION
**Pull request by erinleigh90**
#### Description of changes
Adds category-level customUserAgentDetails to api.graphql, and datastore
* Creates a corresponding private method for API.graphql and GraphQLAPI.graphql that the customer-facing API calls. 
* Requires the use of ts-ignore to allow usage of private methods cross-category.

#### Description of how you validated changes
unit tests

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/api-graphql/src/GraphQLAPI.ts [hunk 1]
```diff
@@ -264,10 +273,11 @@ export class GraphQLAPIClass {
 		}
 	}
 
-	private async _graphql<T = any>(
+	private async _graphqlPromise<T = any>(
```
Comment (erinleigh90) [inline packages/api-graphql/src/GraphQLAPI.ts:276]: Wasn't sure what to rename this to so we can use _graphql as the private method corresponding to graphql. Maybe `_graphqlSend`?

File: packages/datastore/src/sync/processors/sync.ts [hunk 2]
```diff
@@ -213,12 +216,23 @@ class SyncProcessor {
 						this.amplifyConfig
 					);
 
-					return await this.amplifyContext.API.graphql({
-						query,
-						variables,
-						authMode,
-						authToken,
-					});
+					/* TODO: send with actual DataStore action */
```
Comment (stocaaro) [inline packages/datastore/src/sync/processors/sync.ts:219]: Leaving this because actions come through in a future change, right?

**Review by stocaaro — COMMENTED** [review]
Coming together for sure. I have some light comments. Let me know if anything isn't clear.

File: packages/datastore/src/sync/processors/sync.ts (hunk 2 above)
Comment (erinleigh90) [reply packages/datastore/src/sync/processors/sync.ts:219]: Correct

File: packages/core/package.json [hunk 3]
```diff
@@ -52,7 +52,8 @@
 		"find": "^0.2.7",
 		"genversion": "^2.2.0",
 		"prepend-file": "^1.3.1",
-		"react-native": "^0.64.1"
+		"react-native": "^0.64.1",
+		"@aws-sdk/types": "3.6.1"
```
Comment (jimblanc) [inline packages/core/package.json:56]: Seems like this dependency is a duplicate (line 50)

File: packages/datastore/src/sync/processors/subscription.ts [hunk 4]
```diff
@@ -443,16 +445,23 @@ class SubscriptionProcessor {
 											}`
 										);
 
+										// @ts-ignore Use private method to send internal metrics
+										const garphQlResult = this.amplifyContext.API._graphql(
```
Comment (jimblanc) [inline packages/datastore/src/sync/processors/subscription.ts:449]:
```suggestion
										const graphQlResult = this.amplifyContext.API._graphql(
```

File: packages/datastore/src/sync/processors/subscription.ts [hunk 5]
```diff
@@ -443,16 +445,23 @@ class SubscriptionProcessor {
...
 											Observable<{
 												value: GraphQLResult<Record<string, PersistentModel>>;
 											}>
-										>(<unknown>this.amplifyContext.API.graphql({
-											query,
-											variables,
-											...{ authMode },
-											authToken,
-										}));
+										>(<unknown>garphQlResult);
```
Comment (jimblanc) [inline packages/datastore/src/sync/processors/subscription.ts:464]:
```suggestion
										>(<unknown>graphQlResult);
```

File: packages/api-graphql/src/GraphQLAPI.ts (hunk 1 above)
Comment (jimblanc) [reply packages/api-graphql/src/GraphQLAPI.ts]: Maybe `_graphqlInvoke` or `_graphqlAsync` (instead of `Promise` if we want to stay with that pattern).

**Review by jimblanc — COMMENTED** [review]
Looks good, some nits below

File: packages/api-graphql/src/GraphQLAPI.ts (hunk 1 above)
Comment (erinleigh90) [reply packages/api-graphql/src/GraphQLAPI.ts]: I went with _graphqlSend, but invoke works too

**Review by jimblanc — APPROVED** [review]

**Pull request closed without merging**